amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpRpcFlowStage.scala

```diff
@@ -90,10 +92,25 @@ final class AmqpRpcFlowStage(settings: AmqpSinkSettings, bufferSize: Int, respon
               body: Array[Byte]
           ): Unit =
             consumerCallback.invoke(
-              CommittableIncomingMessage(
-                IncomingMessage(ByteString(body), envelope, properties),
-                commitCallback
-              )
+              new CommittableIncomingMessage {
+                override val message = IncomingMessage(ByteString(body), envelope, properties)
+
+                override def ack(multiple: Boolean) = {
+                  Future {
+                    channel.basicAck(message.envelope.getDeliveryTag, multiple)
```

Comment:
Maybe I was a bit unclear, I think this should go into the `commitCallback` logic, so that it is executed on the same thread as the stage itself and the future is not completed until both ack/nack and changing the counter is done, that means we need to pass a promise from here to the async callback, call the channel method, update the counter and then (if successful) complete the promise. 

There will still be some unknowns with that though, what if the stage already failed/stopped when a downstream stage calls ack/nack? What if it was alive when ack/nack was called but fails before the async callback has been executed. There is a solution for that coming up with https://github.com/akka/akka/pull/23185 though, so maybe complete this and then update to use that later.

Reply:
This future uses the `materializer.executionContext` from the stage. Which as far as I understand is exactly how Kafka and IronMQ connectors do it.

The call is blocking so the future won't complete until the call is done.

Adding the call to `channel.ack` or `channel.nack` to the `commitCallback` has the issue that we won't be able to use parameters (or we need a separate callback for each combination of parameters).

Can you elaborate a bit in what exactly you would like me to do here? I'm a bit lost :)

Regarding the fail/stop uncertainties. I think that we should complete this work and then update later as you suggest.

Reply:
The counting down inflight messages and the nack is now done in separate places/threads even though they both belong to the same "logical" operation. The future will can complete even though the `asyncCallback` has not executed yet

Handling the two different calls with a single callback can be done by just encoding the needed data plus the promise in a single or two case classes with a common supertype whose sole purpose is to be passed across the async boundary for the stage to do things with and then complete the promise. (something like how we do it in the `InputStreamStage` for example: https://github.com/akka/akka/blob/3ba093d27e830bea675e45d0cec56860de3d133e/akka-stream/src/main/scala/akka/stream/impl/io/InputStreamSinkStage.scala#L56 )

Reply:
To say it in another way, rather than scheduling two blocks of logic to execute, the `Future` and the `asyncCallback` we can eliminate one, and execute everything in the `asyncCallback`.